test/gtest/ucp/test_ucp_am.cc

```diff
@@ -0,0 +1,290 @@
+                                                    ucp_ep_h reply_ep,
+                                                    unsigned flags)
+{
+    test_ucp_am_base *self = reinterpret_cast<test_ucp_am_base*>(arg);
+    self->replies++;
+    return UCS_OK;
+}
+
+ucs_status_t test_ucp_am_base::ucp_process_am_cb(void *arg, void *data,
+                                                 size_t length,
+                                                 ucp_ep_h reply_ep,
+                                                 unsigned flags)
+{
+    test_ucp_am_base *self = reinterpret_cast<test_ucp_am_base*>(arg);
+    if (reply_ep) {
+        self->reply = ucp_am_send_nb(reply_ep, UCP_REPLY_ID, NULL, 1,
+                                     ucp_dt_make_contig(0),
+                                     ucp_send_am_cb, 0);
+        EXPECT_FALSE(UCS_PTR_IS_ERR(self->reply));
+        if (UCS_PTR_STATUS(self->reply) == UCS_OK) {
+            self->reply = NULL;
+        }
+    }
+    return self->am_handler(self, data, length, flags);
+
+}
+
+ucs_status_t test_ucp_am_base::am_handler(test_ucp_am_base *me, void *data,
+                                          size_t length, unsigned flags)
+{
+    ucs_status_t status;
+    std::vector<char> cmp(length, (char)length);
+    std::vector<char> databuf(length, 'r');
+
+    memcpy(&databuf[0], data, length);
+
+    EXPECT_EQ(cmp, databuf);
+    if (me->release) {
+        me->for_release[me->recv_ams] = data;
+        status                        = UCS_INPROGRESS;
+    } else {
+        status = UCS_OK;
+    }
+    me->recv_ams++;
+    return status;
+}
+
+class test_ucp_am : public test_ucp_am_base
+{
+    public:
+        ucp_ep_params_t get_ep_params() {
+            ucp_ep_params_t params = test_ucp_am_base::get_ep_params();
+            params.field_mask     |= UCP_EP_PARAM_FIELD_FLAGS;
+            params.flags          |= UCP_EP_PARAMS_FLAGS_NO_LOOPBACK;
+            return params;
+        }
+        virtual void init(){
+            ucp_test::init();
+            sender().connect(&receiver(), get_ep_params());
+            receiver().connect(&sender(), get_ep_params());
+        }
+
+    protected:
+        void do_set_am_handler_realloc_test();
+        void do_send_process_data_test(int test_release, uint16_t am_id,
+                                       int send_reply);
+        void do_send_process_data_iov_test();
+        void set_handlers(uint16_t am_id);
+        void set_reply_handlers(uint16_t am_id);
+};
+
+void test_ucp_am::set_reply_handlers(uint16_t am_id)
+{
+    ucp_worker_set_am_handler(sender().worker(), am_id,
+                              ucp_process_reply_cb, this,
+                              UCP_AM_FLAG_WHOLE_MSG);
+    ucp_worker_set_am_handler(receiver().worker(), am_id,
+                              ucp_process_reply_cb, this,
+                              UCP_AM_FLAG_WHOLE_MSG);
+}
+
+void test_ucp_am::set_handlers(uint16_t am_id)
```

Comment:
no need to pass am_id, can use the const def inside

Reply:
I could rewrite it, but I pass in different values.

Reply:
you are right, this is only relevant for set_reply_handlers()